[feat] Integrate audio recording into ChatInput widget

sfc-gh-nbellante opened the pull request:
## Describe your changes

This PR refactors the AudioInput component and adds audio recording capabilities to the ChatInput component:

1. In AudioInput:
   - Improved code organization by moving the transcodeAndUploadFile function before controller initialization
   - Removed the unnecessary ref for transcodeAndUploadFile and passed the function directly to onApprove
   - Added cleanup for blob URLs to prevent memory leaks
   - Added explanatory comments for error handling cases

2. In ChatInput:
   - Added audio recording functionality with mic button when acceptAudio is enabled
   - Implemented recording UI with waveform visualization and approve/cancel buttons
   - Added audio file upload and submission logic
   - Ensured proper cleanup of audio resources
   - Fixed setValue handling to prevent duplicate processing

## Testing Plan

- The AudioInput refactoring maintains the same functionality with improved code organization
- Manually tested audio recording in ChatInput with various browsers
- Verified that audio files are properly uploaded and submitted with chat messages
- Tested error handling for recording and upload failures
- Verified that UI state transitions correctly between text input and recording modes
- Confirmed that audio resources are properly cleaned up when component unmounts

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

sfc-gh-nbellante commented:
* **#12836** <a href="https://app.graphite.dev/github/pr/streamlit/streamlit/12836?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#12835** <a href="https://app.graphite.dev/github/pr/streamlit/streamlit/12835?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/streamlit/streamlit/12835?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#12833** <a href="https://app.graphite.dev/github/pr/streamlit/streamlit/12833?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>: 1 other dependent PR ([#12834](https://github.com/streamlit/streamlit/pull/12834) <a href="https://app.graphite.dev/github/pr/streamlit/streamlit/12834?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>)
* **#12832** <a href="https://app.graphite.dev/github/pr/streamlit/streamlit/12832?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `develop`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

sfc-gh-nbellante commented on frontend/lib/src/components/widgets/AudioInput/AudioInput.tsx:
```diff
@@ -199,6 +144,7 @@ const AudioInput: React.FC<Props> = ({
           }
           currentBlobUrlRef.current = blobUrl
         } catch {
+          // Blob URL creation failed - error state will be set below
```
**praise:** I appreciate these succinct comments that make the code easier to understand and navigate.

sfc-gh-nbellante commented on frontend/lib/src/components/widgets/ChatInput/ChatInput.tsx:
```diff
@@ -324,15 +509,56 @@ function ChatInput({
+      // Approve the recording (encodes to WAV and triggers onApprove event which handles upload)
+      await controller.approve(blob)
+    } catch {
+      // Error handling is done via controller events
```
**question:** If error handling is done by the controller events, why should we need the try/catch or throwing here?

sfc-gh-nbellante commented on frontend/lib/src/components/widgets/ChatInput/ChatInput.tsx:
```diff
@@ -398,8 +636,34 @@ function ChatInput({
+              autoExpand.isExtended || controller.state === "recording"
+            }
+          >
+            {/* Waveform - always mounted to ensure ref is available for initialization */}
```
**suggestion:** We should really avoid inline styles like this unless absolutely necessary.

sfc-gh-nbellante commented on frontend/lib/src/components/widgets/AudioInput/AudioInput.tsx:
```diff
@@ -285,7 +225,65 @@ const AudioInput: React.FC<Props> = ({
+    events: {
+      onPermissionDenied: async () => {
+        setHasNoMicPermissions(true)
+        return Promise.resolve()
```
**complaint:** I'm not a huge fan of all these Promise.resolve()'s.

sfc-gh-nbellante commented on frontend/lib/src/components/widgets/ChatInput/ChatInput.tsx:
```diff
@@ -112,6 +119,68 @@ function ChatInput({
+            fileUrls.uploadUrl as string,
+            audioFile,
+            () => {
+              // Progress callback - not needed for silent upload
```
**question:** Could/should we just make this optional? Alternatively, should we be using this progress callback somewhere in the ui?

sfc-gh-nbellante commented on frontend/lib/src/components/widgets/ChatInput/ChatInput.tsx:
```diff
@@ -112,6 +119,68 @@ function ChatInput({
+            chatInputRef.current.focus()
+          }
+        }
+        /* eslint-enable @typescript-eslint/no-use-before-define */
```
**question:** What's getting used before defined?

sfc-gh-nbellante commented on frontend/lib/src/components/widgets/ChatInput/ChatInput.tsx:
```diff
@@ -142,6 +211,25 @@ function ChatInput({
+        // Fire-and-forget deletion - errors are not critical to user flow
+        uploadClient.deleteFile(file.status.fileUrls.deleteUrl).catch(() => {
+          // Silently ignore deletion errors - file may already be deleted or server unavailable
+          // This is a cleanup operation that shouldn't block the user
```
**suggestion:** Perhaps we should use our error logging to track this?

sfc-gh-nbellante commented on frontend/lib/src/components/widgets/ChatInput/ChatInput.tsx:
```diff
@@ -150,47 +238,36 @@ function ChatInput({
+    }, [files])
 
-  const getNextLocalFileId = (): number => {
+  const getNextLocalFileId = useCallback((): number => {
```
**question:** Is this cheap of a function worth `useCallback`-ing?

sfc-gh-nbellante commented on frontend/lib/src/components/widgets/ChatInput/ChatInput.tsx:
```diff
@@ -383,7 +525,7 @@ function ChatInput({
     <>
       {acceptFile === AcceptFileValue.None ? null : (
-        <ChatUploadedFiles items={[...files]} onDelete={deleteFile} />
+        <ChatUploadedFiles items={files} onDelete={deleteFile} />
```
**question/concern/suggestion:** It looks like the original author was intentionally creating a copy of `files` here and in your changes you removed that. Why? How can we be certain this didn't break anything? Look in the git history to figure out why this was added like this to understand the backstory here.

sfc-gh-nbellante commented on frontend/lib/src/components/widgets/ChatInput/ChatInput.tsx:
```diff
@@ -150,47 +238,120 @@ function ChatInput({
+      name,
+      size,
+    })
+  }, [audioFile])
```
The function createChatInputAudioFileInfo was removed in commit 6f00f3e49e1 as part of a refactoring that simplified the audio submission flow.

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/ChatInput/ChatInput.tsx:
```diff
@@ -79,6 +83,8 @@ export interface Props {
   fragmentId?: string
 }
 
+const LOG = getLogger("ChatInput")
```
**question:** Why not reuse the `LOG` that is defined here `frontend/lib/src/components/ChatInput/logger.ts`?

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/ChatInput/ChatInput.tsx:
```diff
@@ -274,32 +307,129 @@ function ChatInput({
+    events: controllerEvents,
+  })
+
+  const handleSubmit = useCallback((): void => {
+    submitChatInput()
+  }, [submitChatInput])
```
**suggestion:** This seems to be an unnecessary wrapping with `useCallback`, just using `submitChatInput` would have the same effect throughout.

sfc-gh-nbellante replied:
So the first thing it did was a:
`onClick={() => submitChatInput()}`
because `submitChatInput` has an optional first argument (`audioInfo?: UploadedFileInfoProto`)
and doing `onClick={submitChatInput}` won't work because then we're gonna have a click handler event getting shoved into a `UploadedFileInfoProto` sized hole. So this felt like the least bad solution given the first solution would always cause a reallocation+rerender, right? Open to others I didn't think of though.

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/ChatInput/styled-components.ts:
```diff
@@ -115,3 +125,20 @@ export const StyledInputInstructionsContainer = styled.div(({ theme }) => ({
+    alignItems: "center",
+    // Add right padding to account for the two buttons (cancel + approve)
+    // Each button is: iconSizes.xl + 2 * spacing.sm
+    // Total for 2 buttons: 2 * (iconSizes.xl + 2 * spacing.sm) + spacing.sm (gap)
```
**question:** Is there a better way to structure this? This feels like a codesmell to reserve a specific amount of space for items when something like flexbox should be able to do this for us.

sfc-gh-nbellante committed:
Integrate audio recording into ChatInput widget

Add audio recording capability to the ChatInput widget with waveform visualization and approve/cancel controls. Includes shared logger for ChatInput components.

- Add WaveformController integration for live audio recording
- Add recording state management and UI controls
- Add shared logger.ts for ChatInput component logging
- Add comprehensive test coverage for audio recording flow

sfc-gh-nbellante committed:
Refactor AudioInput to use controllable waveform pattern

Update AudioInput and useWaveformController to support being controlled externally by ChatInput. Changes include:

- Add controllerRef pattern to allow parent components to access controller methods
- Fix circular dependency by moving upload logic out of controller events
- Add error handling comments for silent catch blocks
- Update AudioInput test to match new controller pattern

sfc-gh-nbellante committed:
Fix missing styled component after removing ChatInput library

Move StyledChatAudioWave from removed component library to widget's styled-components file and update import path.

sfc-gh-nbellante committed:
Fix ESLint import order errors in ChatInput

- Consolidate styled-components imports into single block
- Add empty line between import groups
- Move StyledChatAudioWave to correct import location

sfc-gh-nbellante committed:
Fix import group formatting - remove empty line within group

sfc-gh-nbellante committed:
Add width to StyledChatAudioWave to fix invisible waveform

When copying StyledChatAudioWave from PR7, we missed the width: 100% that was added in PR8's inline styles refactor. Without explicit width, the waveform collapses to 0 width and becomes invisible despite recording working.

sfc-gh-nbellante committed:
Fix recording layout by using static positioning for buttons

When recording, use static positioning instead of absolute so buttons participate in flex layout alongside waveform. This fixes overlapping layout issues.

sfc-gh-nbellante committed:
Add padding to button container when recording

sfc-gh-nbellante committed:
Fix placeholder text overlap when audio recording enabled

Add acceptAudio prop to StyledInputInstructionsContainer to account for both mic and send buttons when audio is enabled. Doubles the right padding calculation from 1 button to 2 buttons to prevent overlap.

sfc-gh-nbellante committed:
Account for gap between buttons in placeholder text positioning

Add extra spacing.sm to account for the gap between mic and send buttons when calculating right padding for InputInstructions.

sfc-gh-nbellante committed:
Make button gap consistent at 0 for both recording states

sfc-gh-nbellante committed:
Increase placeholder text margin for audio buttons

Double the final margin from spacing.sm to 2*spacing.sm when acceptAudio is true to prevent overlap with mic button.

sfc-gh-nbellante committed:
Fix textarea paddingRight to account for mic button

Apply the same conditional padding logic to the textarea input that we use for InputInstructions - account for 2 buttons when acceptAudio is true.

sfc-gh-nbellante committed:
Reduce textarea paddingRight to just 2 button widths

Remove extra spacing.sm margin - 80px (2 × 40px buttons) is sufficient padding.

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/ChatInput/styled-components.ts:
```diff
@@ -97,21 +99,62 @@ export const StyledSendIconButton = styled.button<StyledSendIconButtonProps>(
+      bottom: "0px",
+      // Calculate the right padding to account for button(s) on the right
+      // Each button is: iconSizes.xl + 2 * spacing.sm
+      // When acceptAudio is true, there are 2 buttons (mic + send) with extra margin
```
**note:** My previous comment about re-structuring this so we aren't unnecessarily coupling this still stands: https://github.com/streamlit/streamlit/pull/12835/#discussion_r2478755722

sfc-gh-nbellante replied:
I have no idea where this is being used actually. I was trying to tweak this earlier and get it to not overlap the placeholder text, but it wasn't changing anything and I found that the actual padding was coming from a paddingRight elsewhere. Let me see if this can be removed entirely.

sfc-gh-nbellante replied:
Ah, gem from claude.

>   The wrapper exists solely to push the character count left to avoid the absolutely positioned buttons.

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/ChatInput/ChatInput.tsx:
```diff
@@ -408,75 +583,137 @@ function ChatInput({
+                  </StyledSendIconButton>
+                  {/* Approve button (✓ icon or spinner during upload) */}
+                  <StyledSendIconButton
+                    onClick={() => {
+                      handleRecordingApprove().catch(_error => {
+                        // Error handling is done via controller events
+                      })
```
**suggestion (non-blocking):** This should be wrapped in a `useCallback`.

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/ChatInput/ChatInput.tsx:
```diff
@@ -408,75 +583,137 @@ function ChatInput({
+                  {/* Mic button */}
+                  {acceptAudio && (
+                    <StyledSendIconButton
+                      onClick={(e: React.MouseEvent) => {
+                        handleMicClick(e).catch(_error => {
+                          // Error handling is done via controller events
+                        })
+                      }}
```
**suggestion (non-blocking):** This should be wrapped in a `useCallback`.

sfc-gh-nbellante commented:
### Merge activity

* **Nov 1, 8:20 AM UTC**: @sfc-gh-nbellante merged this pull request with [Graphite](https://app.graphite.dev/github/pr/streamlit/streamlit/12835).
